### ISSUE TITLE
1.18: kube-apiserver:  fatal error: concurrent map writes

### ISSUE THREAD

**Issue by lance5890** [kubernetes/kubernetes#111943]
### What happened?

apiserver  panic by some requests, and finally it will crash to restart ,and this situation will replay after one day

```bash
2022-08-17T23:23:26.190411273+08:00 stderr F E0817 15:23:26.190027       1 wrap.go:39] apiserver panic'd on GET /api/v1/namespaces/kube-system/pods?continue=eyJ2IjoibWV0YS5rOHMuaW8vdjEiLCJydiI6NDQwMzY4MzI2LCJzdGFydCI6Im92cy12c3dpdGNoZC1nNTk3Mlx1MDAwMCJ9&limit=500
2022-08-17T23:23:27.38027444+08:00 stderr F E0817 15:23:27.380126       1 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"context canceled"}
2022-08-17T23:23:27.385965166+08:00 stderr F E0817 15:23:27.385874       1 writers.go:105] apiserver was unable to write a JSON response: client disconnected
2022-08-17T23:23:27.385993129+08:00 stderr F E0817 15:23:27.385905       1 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"client disconnected"}
2022-08-17T23:23:27.38704366+08:00 stderr F E0817 15:23:27.386985       1 writers.go:118] apiserver was unable to write a fallback JSON response: client disconnected
2022-08-17T23:23:27.984046186+08:00 stderr F E0817 15:23:27.983907       1 writers.go:105] apiserver was unable to write a JSON response: client disconnected
2022-08-17T23:23:27.984074048+08:00 stderr F E0817 15:23:27.983944       1 status.go:71] apiserver received an error that is not an metav1.Status: &errors.errorString{s:"client disconnected"}
2022-08-17T23:23:28.08048774+08:00 stderr F fatal error: concurrent map writes
2022-08-17T23:23:28.088010054+08:00 stderr F 
2022-08-17T23:23:28.088028237+08:00 stderr F goroutine 83957516 [running]:
2022-08-17T23:23:28.088031962+08:00 stderr F runtime.throw(0x464d8f7, 0x15)
2022-08-17T23:23:28.088045506+08:00 stderr F 	/usr/local/go/src/runtime/panic.go:774 +0x72 fp=0xc023dc3940 sp=0xc023dc3910 pc=0x42f902
2022-08-17T23:23:28.088052094+08:00 stderr F runtime.mapassign_faststr(0x4065480, 0xc04ed395f0, 0xc08fa85598, 0x8, 0x4001)
2022-08-17T23:23:28.088055026+08:00 stderr F 	/usr/local/go/src/runtime/map_faststr.go:211 +0x417 fp=0xc023dc39a8 sp=0xc023dc3940 pc=0x413777
2022-08-17T23:23:28.088058726+08:00 stderr F net/textproto.MIMEHeader.Set(...)
2022-08-17T23:23:28.088060909+08:00 stderr F 	/usr/local/go/src/net/textproto/header.go:22
2022-08-17T23:23:28.088064716+08:00 stderr F net/http.Header.Set(...)
2022-08-17T23:23:28.08806739+08:00 stderr F 	/usr/local/go/src/net/http/header.go:37
2022-08-17T23:23:28.088071717+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.(*auditResponseWriter).setHttpHeader(0xc00e3470e0)
2022-08-17T23:23:28.088075832+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit.go:192 +0xec fp=0xc023dc3a10 sp=0xc023dc39a8 pc=0x1501e8c
2022-08-17T23:23:28.08807956+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters.(*auditResponseWriter).Write(0xc00e3470e0, 0xc051016200, 0x61, 0xf8, 0x0, 0x0, 0x0)
2022-08-17T23:23:28.088082981+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/filters/audit.go:212 +0x46 fp=0xc023dc3a58 sp=0xc023dc3a10 pc=0x1501fb6
2022-08-17T23:23:28.088086268+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.(*ResponseWriterDelegator).Write(0xc016072810, 0xc051016200, 0x61, 0xf8, 0x0, 0x0, 0x0)
2022-08-17T23:23:28.088089557+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:449 +0x5b fp=0xc023dc3aa0 sp=0xc023dc3a58 pc=0x14ee7cb
2022-08-17T23:23:28.088092638+08:00 stderr F compress/flate.(*huffmanBitWriter).write(...)
2022-08-17T23:23:28.088096683+08:00 stderr F 	/usr/local/go/src/compress/flate/huffman_bit_writer.go:136
2022-08-17T23:23:28.088099766+08:00 stderr F compress/flate.(*huffmanBitWriter).flush(0xc0510161e0)
2022-08-17T23:23:28.088103438+08:00 stderr F 	/usr/local/go/src/compress/flate/huffman_bit_writer.go:128 +0xe3 fp=0xc023dc3ae8 sp=0xc023dc3aa0 pc=0x6a9063
2022-08-17T23:23:28.088106616+08:00 stderr F compress/flate.(*huffmanBitWriter).writeStoredHeader(0xc0510161e0, 0x0, 0xc023dc3b01)
2022-08-17T23:23:28.088109621+08:00 stderr F 	/usr/local/go/src/compress/flate/huffman_bit_writer.go:412 +0x63 fp=0xc023dc3b10 sp=0xc023dc3ae8 pc=0x6a9f73
2022-08-17T23:23:28.088112812+08:00 stderr F compress/flate.(*compressor).close(0xc0d7962000, 0x70, 0x70)
2022-08-17T23:23:28.088130256+08:00 stderr F 	/usr/local/go/src/compress/flate/deflate.go:647 +0x83 fp=0xc023dc3b38 sp=0xc023dc3b10 pc=0x6a7bc3
2022-08-17T23:23:28.088139135+08:00 stderr F compress/flate.(*Writer).Close(...)
2022-08-17T23:23:28.088141196+08:00 stderr F 	/usr/local/go/src/compress/flate/deflate.go:729
2022-08-17T23:23:28.088143879+08:00 stderr F compress/gzip.(*Writer).Close(0xc02b412000, 0xc09cb1cf00, 0x51bdaa0)
2022-08-17T23:23:28.088146283+08:00 stderr F 	/usr/local/go/src/compress/gzip/gzip.go:242 +0x6a fp=0xc023dc3b88 sp=0xc023dc3b38 pc=0x6b388a
2022-08-17T23:23:28.088148652+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.(*deferredResponseWriter).Close(0xc0ce085950, 0xc049e14e00, 0x70)
2022-08-17T23:23:28.088157991+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:213 +0x7e fp=0xc023dc3bd0 sp=0xc023dc3b88 pc=0x14fa7de
2022-08-17T23:23:28.088168617+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.SerializeObject(0x463922b, 0x10, 0x7f73955cfcc8, 0xc08cb95360, 0x522e0c0, 0xc02df964c0, 0xc0537f9100, 0xc8, 0x51da480, 0xc0b810a990)
2022-08-17T23:23:28.088170461+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:120 +0x417 fp=0xc023dc3c90 sp=0xc023dc3bd0 pc=0x14f9ef7
2022-08-17T23:23:28.088172088+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters.WriteObjectNegotiated(0x5232d00, 0xc01fbab6e0, 0x522e040, 0xc008d604e0, 0xc05092e021, 0xb, 0xc05092e01c, 0x2, 0x522e0c0, 0xc02df964c0, ...)
2022-08-17T23:23:28.088173983+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:251 +0x555 fp=0xc023dc3ef8 sp=0xc023dc3c90 pc=0x14fadc5
2022-08-17T23:23:28.088176498+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.transformResponseObject(0x523db00, 0xc019e04000, 0xc008d604e0, 0xc023dc4758, 0xc0537f9100, 0x522e0c0, 0xc02df964c0, 0xc8, 0x0, 0x0, ...)
2022-08-17T23:23:28.088178106+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/response.go:137 +0x353 fp=0xc023dc4220 sp=0xc023dc3ef8 pc=0x17cb393
2022-08-17T23:23:28.08818298+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers.ListResource.func1(0x522e0c0, 0xc02df964c0, 0xc0537f9100)
2022-08-17T23:23:28.088184892+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/handlers/get.go:276 +0xdfc fp=0xc023dc4870 sp=0xc023dc4220 pc=0x17dd9cc
2022-08-17T23:23:28.088186495+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints.restfulListResource.func1(0xc089df5dd0, 0xc006d2a5b0)
2022-08-17T23:23:28.088190724+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/installer.go:1092 +0x8f fp=0xc023dc48c0 sp=0xc023dc4870 pc=0x17fed0f
2022-08-17T23:23:28.088192413+08:00 stderr F k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics.InstrumentRouteFunc.func1(0xc089df5dd0, 0xc006d2a5b0)
2022-08-17T23:23:28.088194026+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/k8s.io/apiserver/pkg/endpoints/metrics/metrics.go:323 +0x254 fp=0xc023dc4a30 sp=0xc023dc48c0 pc=0x14ef504
2022-08-17T23:23:28.088198675+08:00 stderr F k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).dispatch(0xc000635b00, 0x522e000, 0xc02df964a8, 0xc0537f9100)
2022-08-17T23:23:28.088205684+08:00 stderr F 	/home/jenkins/agent/workspace/CDP_AUTO_TASK/0/2022-04/1475348220813107201_13_140_1519256998197911553/k8s_apiserver/_output/local/go/src/k8s.io/kubernetes/vendor/github.com/emicklei/go-restful/container.go:288 +0xa4f fp=0xc023dc4c08 sp=0xc023dc4a30 pc=0x14d461f
2022-08-17T23:23:28.088212214+08:00 stderr F k8s.io/kubernetes/vendor/github.com/emicklei/go-restful.(*Container).Dispatch(...)

```

### What did you expect to happen?

kube-apiserver do not restart 

### How can we reproduce it (as minimally and precisely as possible)?

this happends often in my environment，but not in other environment, i just do not known how to reproduce it 

### Kubernetes version

<details>

```console
$ kubectl version
# paste output here
k8s 1.18.8
```

</details>

### Cloud provider

<details>
no
</details>

### OS version

<details>

```console
# On Linux:
$ cat /etc/os-release
# paste output here
$ uname -a
# paste output here

# On Windows:
C:\> wmic os get Caption, Version, BuildNumber, OSArchitecture
# paste output here
```

</details>

### Install tools

<details>
kubeadm
</details>

### Container runtime (CRI) and version (if applicable)

<details>

</details>

### Related plugins (CNI, CSI, ...) and versions (if applicable)

<details>

</details>

**Comment by aojea** [comment]
> k8s 1.18.8

that version is EOL, you need to reproduce with a supported version

**Comment by liggitt** [comment]
Is this reproducible on a supported version? Several issues like this were resolved in 1.22+ in https://github.com/kubernetes/kubernetes/issues/107463

**Comment by lance5890** [comment]
yeah, it may be related to #107463 ， but i also wonder why "apiserver was unable to write a JSON response: client disconnected" happens too much , it is caused by apiserver can not response or the client has closed the connection 

> Is this reproducible on a supported version? Several issues like this were resolved in 1.22.6+ in #107463

**Comment by BenTheElder** [comment]
/remove-sig k8s-infra

k8s infra does not own this. https://github.com/kubernetes/community/tree/master/sig-k8s-infra

> Is this reproducible on a supported version?

This, if not, we're going to close this.
We need evidence that this is broken on a currently supported version, 1.18 is way out of support. https://kubernetes.io/releases/

**Comment by liggitt** [comment]
What looks like the same issue was just fixed at HEAD in https://github.com/kubernetes/kubernetes/pull/115282 (reported in https://github.com/kubernetes/kubernetes/issues/122940), to be included in 1.30

/close
as duplicate of https://github.com/kubernetes/kubernetes/issues/122940
